extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/WriteMongoP.java

```diff
@@ -0,0 +1,627 @@
+
+                            BulkWriteResult result = collection.bulkWrite(clientSession, writes);
+
+                            if (!result.wasAcknowledged()) {
+                                commitRetryTracker.attemptFailed();
+                                break;
+                            }
+                        }
+                        clientSession.commitTransaction();
+                        commitRetryTracker.reset();
+                        success = true;
+                    } catch (MongoCommandException e) {
+                        tryExternalRollback();
+                        commitRetryTracker.attemptFailed();
+                        if (!commitRetryTracker.shouldTryAgain() || e.getErrorCode() != MONGODB_TRANSIENT_ERROR) {
+                            throw e;
+                        }
+                    } catch (MongoBulkWriteException e) {
+                        tryExternalRollback();
+                        throw new JetException(e);
+                    } catch (MongoSocketException e) {
+                        tryExternalRollback();
+                        commitRetryTracker.attemptFailed();
+                        if (!commitRetryTracker.shouldTryAgain()) {
+                            throw e;
+                        }
+                    }
+                }
+            }
+        }
+
+        private void startTransactionQuietly() {
+            try {
+                clientSession.startTransaction(TRANSACTION_OPTIONS);
+            } catch (IllegalStateException e) {
+                if (!e.getMessage().contains("already in progress")) {
+                    throw sneakyThrow(e);
+                }
+            }
+        }
+
+        @Override
+        public void rollback() {
+            if (transactionId != null) {
+                tryExternalRollback();
+                documents.clear();
+            }
+        }
+
+        private void tryExternalRollback() {
+            try {
+                clientSession.abortTransaction();
+            } catch (IllegalStateException e) {
+                ignore(e);
+            }
+        }
+
+        @Override
+        public void release() {
+            if (clientSession != null) {
+                clientSession.close();
+            }
+            activeTransactions.remove(transactionId);
+        }
+    }
+
+    private interface CollectionPicker<I> {
+        MongoCollectionKey pick(I item);
+    }
+
+    private final class ConstantCollectionPicker implements CollectionPicker<I> {
+
+        private final String databaseName;
+        private final String collectionName;
+
+        private ConstantCollectionPicker(String databaseName, String collectionName) {
+            this.databaseName = databaseName;
+            this.collectionName = collectionName;
+        }
+
+
+        @Override
+        public MongoCollectionKey pick(I item) {
+            return new MongoCollectionKey(databaseName, collectionName);
+        }
+
+    }
```

Comment:
Maybe remove it and use this?
```java
new FunctionalCollectionPicker(i -> databaseName, i -> collectionName);
```

Reply:
Actually it was other problem; the return value should be cached, that was my idea (avoiding unnecessary object creation and pressure on GC)